### PR TITLE
Fix #12555: Error "grant not found for key id"

### LINKED ISSUE #12555: aws_kms_grant: Error "grant not found for key id" during terraform apply on refresh state
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version
* Terraform v0.12.20
* Terraform AWS provider v2.54

### Affected Resource(s)

* aws_kms_grant

### Terraform Configuration Files

```hcl
resource "aws_kms_key" "test" {
  description = "Test key for grant not found issue"
}

resource "aws_iam_role" "test" {
  name = "test-kms-grant-not-found-role"

  assume_role_policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Action": "sts:AssumeRole",
      "Principal": {
        "Service": "lambda.amazonaws.com"
      },
      "Effect": "Allow",
      "Sid": ""
    }
  ]
}
EOF
}

resource "aws_kms_grant" "test" {
  name              = "test-kms-grant-not-found"
  key_id            = aws_kms_key.test.key_id
  grantee_principal = aws_iam_role.test.arn
  operations        = ["Encrypt", "Decrypt", "GenerateDataKey"]
}
```

### Debug Output
N/A

### Panic Output
N/A

### Expected Behavior
When a KMS grant is not found, it's treated as revoked. Terraform shows the change in the plan and recreates it during the apply.

### Actual Behavior
When a KMS grant was not found, `terraform apply` failed during refreshing state with an error:
```
aws_kms_key.test: Refreshing state... [id=REDACTED_KEY_ID]
aws_iam_role.test: Refreshing state... [id=test-kms-grant-not-found-role]
aws_kms_grant.test: Refreshing state... [id=REDACTED_KEY_ID:REDACTED_GRANT_ID]

Error: grant REDACTED_GRANT_ID not found for key id: REDACTED_KEY_ID
```

### Steps to Reproduce

1. `terraform init`
2. `terraform apply`
3. Revoke the grant created during apply, e.g. with AWS CLI `aws kms revoke-grant --key-id <KEY_ID> --grant-id <GRANT_ID>`
3. `terraform apply`

### Important Factoids
N/A

### References
N/A

### PR DESCRIPTION
**Pull request by dshkyra**
When a KMS grant was not found, terraform apply failed during
refreshing state with an error "grant not found for key id".

Now, when a KMS grant is not found, it is treated as revoked and
terraform shows the change in the plan and recreates it during apply.

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #12555

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource/aws_kms_grant: Fix error "grant not found for key id" during terraform apply during refreshing state
```

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSKmsGrant'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSKmsGrant -timeout 120m
...
--- PASS: TestAccAWSKmsGrant_ARN (38.75s)
--- PASS: TestAccAWSKmsGrant_Basic (39.18s)
--- PASS: TestAccAWSKmsGrant_withRetiringPrincipal (39.32s)
--- PASS: TestAccAWSKmsGrant_bare (40.15s)
--- PASS: TestAccAWSKmsGrant_withConstraints (46.74s)
--- PASS: TestAccAWSKmsGrant_disappears (219.23s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	220.491s
```

### CODE REVIEW

**Comment by bflad** [comment]
☝️ Fixed simple merge conflict since we just merged another PR affecting that file.

File: aws/resource_aws_kms_grant_test.go [hunk 1]
```diff
@@ -168,6 +170,26 @@ func TestAccAWSKmsGrant_ARN(t *testing.T) {
 	})
 }
 
+func TestAccAWSKmsGrant_disappers(t *testing.T) {
```
Comment (bflad) [inline aws/resource_aws_kms_grant_test.go]: Typo 😎 

```suggestion
func TestAccAWSKmsGrant_disappears(t *testing.T) {
```

File: aws/resource_aws_kms_grant_test.go [hunk 2]
```diff
@@ -168,6 +170,26 @@ func TestAccAWSKmsGrant_ARN(t *testing.T) {
...
+			{
+				Config: testAccAWSKmsGrant_Basic("basic", timestamp, "\"Encrypt\", \"Decrypt\""),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSKmsGrantExists("aws_kms_grant.basic"),
+					testAccCheckAWSKmsGrantDisappears("aws_kms_grant.basic"),
+				),
+				ExpectNonEmptyPlan: true,
+			},
+		},
+	})
```
Comment (bflad) [inline aws/resource_aws_kms_grant_test.go]: Due to the merge of another pull request, the input parameters for `testAccAWSKmsGrant_Basic` changed to remove the timestamp. Can you please update to the following?

```suggestion
	resourceName := "aws_kms_grant.test"
	rName := acctest.RandomWithPrefix("tf-acc-test")

	resource.ParallelTest(t, resource.TestCase{
		PreCheck:     func() { testAccPreCheck(t) },
		Providers:    testAccProviders,
		CheckDestroy: testAccCheckAWSKmsGrantDestroy,
		Steps: []resource.TestStep{
			{
				Config: testAccAWSKmsGrant_Basic(rName, "\"Encrypt\", \"Decrypt\""),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckAWSKmsGrantExists(resourceName),
					testAccCheckAWSKmsGrantDisappears(resourceName),
				),
				ExpectNonEmptyPlan: true,
			},
		},
	})
```

File: aws/resource_aws_kms_grant_test.go (hunk 1 above)
Comment (dshkyra) [reply aws/resource_aws_kms_grant_test.go]: A good catch. Thanks!

File: aws/resource_aws_kms_grant_test.go (hunk 2 above)
Comment (dshkyra) [reply aws/resource_aws_kms_grant_test.go]: Updated. Thanks for your code suggestion. I haven't used this nice GitHub "Add suggest to batch" and "Commit suggestion" features before. Pretty convenient.

**Comment by dshkyra** [comment]
@bflad thanks for your comments. I've updated the PR and re-run integration tests :white_check_mark:.

**Pull request merged**